test/integration/mongo_client.test.ts

```diff
@@ -0,0 +1,68 @@
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+import { MongoClient } from '../../src';
+
+describe('MongoClient', () => {
+  let client: MongoClient;
+  let topologyOpenEvents;
+
+  /** Keep track number of call to client connect to close as many as connect (otherwise leak_checker hook will failed) */
+  let clientConnectCounter: number;
+
+  /**
+   * Wrap the connect method of the client to keep track
+   * of number of times connect is called
+   */
+  function clientConnect() {
+    if (!client) {
+      return;
+    }
+    clientConnectCounter++;
+    return client.connect();
+  }
+
+  beforeEach(async function () {
+    client = this.configuration.newClient();
+    topologyOpenEvents = [];
+    clientConnectCounter = 0;
+    client.on('open', event => topologyOpenEvents.push(event));
+  });
+
+  afterEach(async function () {
+    /** Close as many times as connect calls in the runned test (tracked by clientConnectCounter) */
+    const clientClosePromises = [...new Array(clientConnectCounter)].map(() => client.close());
+    await Promise.all(clientClosePromises);
```

Comment:
Appreciate the reference! I understand why we have to close more than once because of the leak checker, sorry about that limitation. 😅  

I'm wondering if `.close` also has a parallelism issue. Supposedly we could be calling `.end` on sockets more than once because as the first "`client.close`" promise is pending the I/O for socket closure the next promise could also synchronously reach the `.end` invocation. 

This may in fact _not_ be the case but _if_ it is we'll want to also fix that in a follow up ticket perhaps.  

I contrived an example, I have to use sleep 1ms to make the promises offset by a bit but I would imagine in real world scenario a delay like that could occur depending on what else is in the event loop.

<details><summary>Example script that calls .end in parallel</summary>

```js
import { once } from 'events';
import { Socket } from 'net';
import { setTimeout as sleep } from 'timers/promises';
import { promisify } from 'util';

const socket = new Socket();
const willConnect = once(socket, 'connect');
socket.connect({ host: '127.0.0.1', port: 2390 }).setDefaultEncoding('utf8').setEncoding('utf8');
await willConnect;

await promisify(callback => socket.write('hello\n', callback))();

// MongoClient .close in parallel
await Promise.all([
  promisify(callback => socket.end(callback))(),
  sleep(1).then(() => promisify(callback => socket.end(callback))())
]);
```

```txt
Error [ERR_STREAM_ALREADY_FINISHED]: Cannot call end after a stream was finished
    at new NodeError (node:internal/errors:393:5)
    at Writable.end (node:internal/streams/writable:637:11)
    at Socket.end (node:net:691:31)
  code: 'ERR_STREAM_ALREADY_FINISHED'
}
```
</details>

Reply:
https://jira.mongodb.org/browse/NODE-5138